src/tools/wasm-opt.cpp

```diff
@@ -253,7 +253,8 @@ int main(int argc, const char* argv[]) {
     auto runPasses = [&]() {
       options.runPasses(*curr);
       if (options.passOptions.validate) {
-        bool valid = WasmValidator().validate(*curr, options.getFeatures());
+        bool valid =
+            WasmValidator().validate(*curr, options.getFeatures(*curr));
```

Comment:
Can be a PassOptions method to read the features, or something like that, and called soon after ModuleReader finishes?

Reply:
(in wasm-opt.cpp, and others can be similar)

Reply:
My rationale for doing it this way is that it makes it harder for users to forget to read the target features section (and if passOptions was private it would be impossible to forget). I also avoid the overhead of reading the target features section more than once by setting and checking `featuresInitialized`. If you still think the code would be more readable with a stateful API that requires users to remember to read the target features section as a separate step, I would be willing to reconsider.

Reply:
Good idea to focus on making it hard to forget reading the target features. One option might be to integrate this in `ModuleReader`. But I'm not sure if there would be cases where a wasm is loaded but we don't care about the features?

If we always care about the features, then that seems like an argument for putting the features on the Module object itself and not on PassOptions. That is, they are part of the Module, really; unlike other stuff on PassOptions like whether we ignore implicit traps - maybe?

Reply:
Yeah it's kind of weird because the module itself encodes a feature set, but not necessarily the one we want to use in passes. That's why I agree with you that it is not great for the module itself to expose the features. I think PassOptions is the best place to keep the features because they are meant to affect the behavior of the passes just like the rest of the pass options. I don't think there should be any cases where we need the PassOption's features but we don't have a module or don't want to consider the module's feature section.